File: app/src/main/java/lizka/reminder/MainActivity.java
package lizka.reminder;

import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentManager;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Toast;

import lizka.reminder.adapter.TabAdapter;
import lizka.reminder.dialog.AddingTaskDialogFragment;
import lizka.reminder.fragment.CurrentTaskFragment;
import lizka.reminder.fragment.DoneTaskFragment;
import lizka.reminder.fragment.SplashFragment;
import lizka.reminder.model.ModelTask;


public class MainActivity extends AppCompatActivity implements AddingTaskDialogFragment.AddingTaskListener {

    FragmentManager fragmentManager;
    PreferenceHelper preferenceHelper;
    TabAdapter tabAdapter;
    CurrentTaskFragment currentTaskFragment;
    DoneTaskFragment doneTaskFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        PreferenceHelper.getInstance().init(getApplicationContext());
        preferenceHelper = PreferenceHelper.getInstance();

        fragmentManager = getSupportFragmentManager();

        runSplash();
        setUI();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action if it is present
        getMenuInflater().inflate(R.menu.menu_main, menu);
        MenuItem splashItem = menu.findItem(R.id.action_splash);
        splashItem.setChecked(preferenceHelper.getBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE));
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();

        // меняем значение по нажатию
        if (id == R.id.action_splash){
            item.setChecked(!item.isChecked());
            preferenceHelper.putBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE, item.isChecked());
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    // метод запуска splashscreen
    public void runSplash() {

        if (!preferenceHelper.getBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE)) {

            SplashFragment splashFragment = new SplashFragment();

            // отображение фрагмента
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, splashFragment)
                    .addToBackStack(null)
                    .commit();

        }
    }

    private void setUI(){
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        if (toolbar != null){
            toolbar.setTitleTextColor(getColor(R.color.white));
            setSupportActionBar(toolbar);
        }

        TabLayout tabLayout = (TabLayout) findViewById(R.id.tab_layout);
        tabLayout.addTab(tabLayout.newTab().setText(R.string.current_task));
        tabLayout.addTab(tabLayout.newTab().setText(R.string.done_task));

        final ViewPager viewPager = (ViewPager) findViewById(R.id.pager);
        tabAdapter = new TabAdapter(fragmentManager, 2);

        viewPager.setAdapter(tabAdapter);
        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));

        tabLayout.setOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                viewPager.setCurrentItem(tab.getPosition());
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });

        currentTaskFragment = (CurrentTaskFragment) tabAdapter.getItem(TabAdapter.CURRENT_TASK_FRAGMENT_POSITION);
        doneTaskFragment = (DoneTaskFragment) tabAdapter.getItem(TabAdapter.DONE_TASK_FRAGMENT_POSITION);

        // вешаем на него слушателя
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DialogFragment addingTaskDialogFragment = new AddingTaskDialogFragment();
                addingTaskDialogFragment.show(fragmentManager, "AddingTaskDialogFragment");
            }
        });
    }

    @Override
    public void onTaskAdded(ModelTask newTask) {
        currentTaskFragment.addTask(newTask);
    }

    @Override
    public void onTaskAddingCancel() {
        Toast.makeText(this, "Task adding cancel", Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/lizka/reminder/MainActivity.kt
package lizka.reminder

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.FragmentManager
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import lizka.reminder.adapter.TabAdapter
import lizka.reminder.db.dao.TaskDao
import lizka.reminder.dialog.AddingTaskDialogFragment
import lizka.reminder.fragment.CurrentTaskFragment
import lizka.reminder.fragment.DoneTaskFragment
import lizka.reminder.fragment.SplashFragment
import lizka.reminder.model.ModelTask
import lizka.reminder.model.ModelTask.STATUS_DONE
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(), AddingTaskDialogFragment.AddingTaskListener {


    lateinit var daoInterface:TaskDao
    lateinit var application: ReminderApplication
    lateinit var fragmentManager: FragmentManager

    lateinit var preferenceHelper:  PreferenceHelper
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        PreferenceHelper.getInstance().init(applicationContext)
        preferenceHelper = PreferenceHelper.getInstance()
        application = getApplication() as ReminderApplication
        daoInterface = application.db.taskDao()
        fragmentManager = supportFragmentManager

        runSplash()

        setUI()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action if it is present
        menuInflater.inflate(R.menu.menu_main, menu)
        val splashItem = menu.findItem(R.id.action_splash)
        splashItem.isChecked = preferenceHelper.getBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        val id = item.itemId

        // меняем значение по нажатию
        if (id == R.id.action_splash) {
            item.isChecked = !item.isChecked
            preferenceHelper.putBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE, item.isChecked)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    // метод запуска splashscreen
    private fun runSplash() {

        if (!preferenceHelper.getBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE)) {

            val splashFragment = SplashFragment()

            // отображение фрагмента
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, splashFragment)
                    .addToBackStack(null)
                    .commit()

        }
    }

    private lateinit var tabAdapter: TabAdapter

    private lateinit var currentTaskFragment: CurrentTaskFragment

    private lateinit var doneTaskFragment: DoneTaskFragment

    private fun setUI() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        if (toolbar != null) {
            toolbar.setTitleTextColor(getColor(R.color.white))
            setSupportActionBar(toolbar)
        }

        val tabLayout = findViewById<TabLayout>(R.id.tab_layout)
        tabLayout.addTab(tabLayout.newTab().setText(R.string.current_task))
        tabLayout.addTab(tabLayout.newTab().setText(R.string.done_task))

        val viewPager = findViewById<ViewPager>(R.id.pager)
        tabAdapter = TabAdapter(fragmentManager, 2)

        viewPager.adapter = tabAdapter
        viewPager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tabLayout))

        tabLayout.setOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                viewPager.currentItem = tab.position
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}

            override fun onTabReselected(tab: TabLayout.Tab) {}
        })

        currentTaskFragment = tabAdapter.getItem(TabAdapter.CURRENT_TASK_FRAGMENT_POSITION) as CurrentTaskFragment
        doneTaskFragment = tabAdapter.getItem(TabAdapter.DONE_TASK_FRAGMENT_POSITION) as DoneTaskFragment

        // вешаем на него слушателя

        fab.setOnClickListener {
            val addingTaskDialogFragment = AddingTaskDialogFragment()
            addingTaskDialogFragment.show(fragmentManager, "AddingTaskDialogFragment")
        }
    }

    override fun createdTask(newTask: ModelTask) {

        Log.i("BUGG", "title = ${newTask.title}")
        val notifyIntent = Intent(this, MyReceiver::class.java)
        notifyIntent.putExtra("taskText", newTask.title)

        val pendingIntent = PendingIntent.getBroadcast(this, 0, notifyIntent, 0)
        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager
        newTask.date
        alarmManager.set(AlarmManager.RTC_WAKEUP, newTask.time, pendingIntent)
        val dateString = SimpleDateFormat("dd/MM/yyyy").format(Date(newTask.time))
        val time = SimpleDateFormat("HH:mm:ss").format(Date(newTask.time))
        Log.i("BUGG", "data: = $dateString" + "time = $time")
        addCurrentTask(newTask)
    }

    override fun onTaskAddingCancel() {}

    fun addDoneTask(task: ModelTask) {
        doneTaskFragment.addTask(task)
        GlobalScope.launch(Dispatchers.Main){
            withContext(Dispatchers.IO){
                daoInterface.delete(task)
                daoInterface.insert(task)
            }
        }
        currentTaskFragment.addTask(task)

    }

    fun addCurrentTask(task: ModelTask) {
        GlobalScope.launch(Dispatchers.Main){
            withContext(Dispatchers.IO){
                daoInterface.delete(task)
                daoInterface.insert(task)
            }
        }
        currentTaskFragment.addTask(task)
    }

    fun removeTask(task: ModelTask) {
        GlobalScope.launch(Dispatchers.Main){
            withContext(Dispatchers.IO){
                daoInterface.delete(task)
            }
        }
    }

    fun getAllCurrent(callBackUI: (MutableList<ModelTask>) -> Unit){
        GlobalScope.launch(Dispatchers.Main){
            val allList = getAll()
            val filteredList = allList.filter { it.status != STATUS_DONE}.toMutableList()
            callBackUI(filteredList)
        }
    }

    fun getAllDone(callBackUI: (MutableList<ModelTask>) -> Unit){
        GlobalScope.launch(Dispatchers.Main){
            val allList = getAll()
            val filteredList = allList.filter { it.status == STATUS_DONE}.toMutableList()
            callBackUI(filteredList)
        }
    }

    private suspend fun getAll():List<ModelTask> = withContext(Dispatchers.IO){
        daoInterface.getAll()
    }
}